lib/report_plugins/daytoday.js

```diff
@@ -283,6 +294,110 @@ daytoday.report = function report_daytoday(datastorage,sorteddaystoshow,options)
           return sel;
       }
 
+    // PREDICTIONS START
+    //
+    function preparePredictedData() {
+
+      var treatmentsTimestamps = []; // Only timestamps for (carbs and bolus insulin) treatments will be captured in this array
+      treatmentsTimestamps.push(dataRange[0]); // Create a fake timestamp at midnight so we can show predictions during night
+      for (var i in data.treatments) {
+        var treatment = data.treatments[i];
+        if (undefined != treatment.carbs && null != treatment.carbs && treatment.carbs > 0) {
+          if (treatment.timestamp)
+            treatmentsTimestamps.push(treatment.timestamp);
+          else if (treatment.created_at)
+            treatmentsTimestamps.push(treatment.created_at);
+        }
+        if (undefined != treatment.insulin && null != treatment.insulin && treatment.insulin > 0) {
+          if (treatment.timestamp)
+            treatmentsTimestamps.push(treatment.timestamp);
+        else if (treatment.created_at)
+          treatmentsTimestamps.push(treatment.created_at);
+        }
+      }
+
+      var predictions = [];
+      if (data && data.devicestatus) {
+        for (var i = data.devicestatus.length - 1; i >= 0; i--) {
+          if (data.devicestatus[i].loop && data.devicestatus[i].loop.predicted) {
+            predictions.push(data.devicestatus[i].loop.predicted);
+          } else if (data.devicestatus[i].openaps && data.devicestatus[i].openaps.suggested && data.devicestatus[i].openaps.suggested.predBGs) {
+            var entry = {};
+            entry.startDate = data.devicestatus[i].openaps.suggested.timestamp;
+            entry.values = data.devicestatus[i].openaps.suggested.predBGs.IOB;
```

Comment:
I'm not sure that's always the correct prediction curve to show for openaps, can we also show the cob curve and maybe the others to?

Reply:
Hello Alimhassam

There is a checkbox to *not* truncate the prediction at the next treatment found. But I can think that with OpenAPS SMB it would produce very many prediction lines after a meal.

It would perhaps make sense to add a time range; if a new treatment is found within the range, say 2 hours, the prediction is not truncated but runs past this treatment. That would skip over the many SMB treatments. But then you would also need a way to advance to the next SMB treatment.

As I'm not using OpenAPS myself I don't know for certain what the data looks like (have been examining Loop and AndroidAPS Nightscout entries) but perhaps the SMB entries are marked as SMB and could be treated specially.

Or in the first version it could perhaps focus on Loop, AndroidAPS and OpenAPS without SMB.

I don't think the COB curve is saved in the predictions data so not possible to display future values. But for past values Nightscout Day to day already has option to show IOB and COB.

Reply:
OpenAps generates a couple of different prediction curves. IOB, COB, UAM, ZT. IOB. I'm not sure that iob is always the best one to display. 
From my understanding, IOB is the prediction curve taking iob only into account. COB takes uses both iob and cob, *but is only available when cob are present* (maybe that's why you didn't see it?). UAM is the unannounced meal prediction, and ZT is the prediction if carb absorbtion stops, and we use a zero temp basal. (cc @scottleibrand correct me if i'm wrong). 
Ideally i think it might be nice to show all prediction curves, if it's possible, i'm not sure showing the iob one is always the right choice. 

example device status record:

```
{
    "_id": {
        "$oid": "5a41acc0bae4c2001345d309"
    },
    "device": "...",
    "openaps": {
        "iob": {
            "iob": 1.373,
            "activity": 0.0025,
            "basaliob": -0.269,
            "bolusiob": 1.641,
            "netbasalinsulin": -3.05,
            "bolusinsulin": 5.8,
            "iobWithZeroTemp": {
                "iob": 1.373,
                "activity": 0.0025,
                "basaliob": -0.269,
                "bolusiob": 1.641,
                "netbasalinsulin": -3.05,
                "bolusinsulin": 5.8,
                "time": "2017-12-26T01:56:53.000Z"
            },
            "lastBolusTime": 1514253349000,
            "lastTemp": {
                "rate": 1,
                "timestamp": "2017-12-25T20:55:06-05:00",
                "started_at": "2017-12-26T01:55:06.000Z",
                "date": 1514253306000,
                "duration": 3.33
            },
            "timestamp": "2017-12-26T01:56:53.000Z"
        },
        "suggested": {
            "temp": "absolute",
            "bg": 89,
            "tick": "+4",
            "eventualBG": 300,
            "insulinReq": 1.9,
            "reservoir": "106.9",
            "deliverAt": "2017-12-26T01:57:30.669Z",
            "sensitivityRatio": 1.11,
            "predBGs": {
                "IOB": [
                    89,
                    92,
                    94,
                    96,
                    97,
                    97,
                    96,
                    95,
                    94,
                    92,
                    90,
                    87,
                    84,
                    82,
                    79,
                    76,
                    73,
                    71,
                    68,
                    66,
                    64,
                    61,
                    59,
                    57,
                    55,
                    54,
                    52,
                    50,
                    49,
                    47,
                    46,
                    45,
                    44,
                    43,
                    42,
                    41,
                    40,
                    39
                ],
                "ZT": [
                    89,
                    88,
                    87,
                    85,
                    84,
                    82,
                    80,
                    78,
                    76,
                    74,
                    73,
                    71,
                    70,
                    70,
                    69,
                    69,
                    69,
                    70,
                    70,
                    71,
                    73,
                    75,
                    76,
                    79,
                    81,
                    84,
                    87,
                    90,
                    94,
                    97,
                    101
                ],
                "COB": [
                    89,
                    93,
                    97,
                    101,
                    106,
                    110,
                    115,
                    121,
                    126,
                    133,
                    139,
                    147,
                    155,
                    163,
                    172,
                    182,
                    192,
                    203,
                    215,
                    226,
                    236,
                    245,
                    254,
                    263,
                    270,
                    277,
                    283,
                    288,
                    293,
                    297,
                    301,
                    303,
                    305,
                    307,
                    307,
                    307,
                    306,
                    305,
                    305,
                    304,
                    303,
                    303,
                    302,
                    302,
                    301,
                    301,
                    300
                ],
                "UAM": [
                    89,
                    92,
                    95,
                    96,
                    98,
                    98,
                    98,
                    98,
                    97,
                    96,
                    95,
                    94,
                    92,
                    90,
                    88,
                    85,
                    83,
                    80,
                    78,
                    75,
                    73,
                    71,
                    69,
                    67,
                    65,
                    63,
                    62,
                    60,
                    58,
                    57,
                    56,
                    54,
                    53,
                    52,
                    51,
                    50,
                    49,
                    49,
                    48,
                    47,
                    47,
                    46,
                    45,
                    45,
                    45,
                    44,
                    44,
                    43
                ]
            },
            "COB": 45,
            "IOB": 1.373,
            "reason": "COB: 45, Dev: 1.4, BGI: 0.0, ISF: 3.3, CR: 9.07, Target: 5.7, minPredBG 11.9, minGuardBG 5.1, IOBpredBG 2.2, COBpredBG 16.7, UAMpredBG 2.4;  insulinReq 1.9; setting 60m low temp of 0U/h. Waiting 1m 18s to microbolus again. ",
            "rate": 0,
            "duration": 60,
            "timestamp": "2017-12-26T01:57:30.815Z"
        },
        "enacted": {
            "insulinReq": 2.43,
            "bg": 85,
            "reservoir": "108.2",
            "temp": "absolute",
            "rate": 1,
            "reason": "COB: 46, Dev: 0.2, BGI: 0.0, ISF: 3.3, CR: 9.07, Target: 5.7, minPredBG 13.7, minGuardBG 4.8, IOBpredBG 4.3, COBpredBG 20.7, UAMpredBG 4.2;  insulinReq 2.43; setting 30m low temp of 0.99U/h. Microbolusing 1.2U. ",
            "IOB": 0.177,
            "duration": 30,
            "sensitivityRatio": 1.11,
            "COB": 46,
            "eventualBG": 372,
            "recieved": true,
            "units": 1.2,
            "tick": "+0",
            "timestamp": "2017-12-26T01:55:05.244Z",
            "deliverAt": "2017-12-26T01:54:53.762Z"
        }
    },
    "pump": {
        "clock": "2017-12-25T20:56:53-05:00",
        "battery": {
            "status": "normal",
            "voltage": 1.48
        },
        "reservoir": 106.9,
        "status": {
            "status": "normal",
            "bolusing": false,
            "suspended": false,
            "timestamp": "2017-12-26T01:57:13.625Z"
        }
    },
    "uploader": {
        "batteryVoltage": 3662,
        "battery": 47
    },
    "created_at": "2017-12-26T01:58:24.814Z"
}
```

Reply:
also, I just noticed the truncate prediction button, you're right, that it makes the graph quite crowded, with smb, but it's useful to have that option, thanks!

Reply:
@alimhassam I plotted the data you supplied and it looks like COB curve shows the predicted BG only taking carbs (so no IOB) into account. IOB curve shows the predicted BG taking both carbs and IOB into account.

![pred1](https://user-images.githubusercontent.com/7753218/34469616-91c0d2a2-ef23-11e7-81a1-333be9073dfa.jpg)

Predictions could plot all curves (checkboxes to select which you would like to see), but I think that for most people and situations the IOB curve is actually the most relevant to plot.


This is what the Predictions may look like for a Loop user. I have already benefited from it b/c I have seen that BG for my son rises much quicker than what Loop expects. 

![pred2](https://user-images.githubusercontent.com/7753218/34469617-976238ae-ef23-11e7-8129-c18de62550d5.jpg)

Reply:
@lixgbg no, you have it backwards: COB takes into account both IOB and COB, not just COB.  IOB doesn't take carbs into account.

What's the logic for deciding which predictions to show?  The dynamic carb absorption logic in both Loop and OpenAPS cause the predictions to be updated with each new CGM data point, so taking just the prediction immediately after each new carb entry doesn't necessarily provide the most relevant predictions...

Reply:
@scottleibrand Ok, then it's the COB curve that would make sense to plot.

The whole purpose of the Predictions feature is to be able to see how well the prediction made at the time of eating carbs and/or making a meal bolus treatment matches the actual outcome and using this insight be able to select the best carb absorption profile (Loop has e.g. 2h, 3h, and 4h as default) and help tune the ISF, IC etc. And then, ultimately (future), be able to provide input to new personalized carb absorption profiles in Closed Loop systems.

At the time of eating carbs & bolusing the closed loop system makes a decision on how much bolus to suggest. After that bolus is given any temp basal adjustments are just minor corrections up/down to counteract for the real outcome. So the better the initial prediction can be made the higher chance of a good BG response.

OpenAPS users using AMA/UAM/SMB may not be the primary targets for the NS Predictions feature as they have an even more dynamic system than Loop and hence depend less on the initial prediction.

Perhaps the Predictions enable checkbox on Day to day reports should be renamed "Loop", as there is already an OpenAPS checkbox?

Reply:
Ok, that makes sense.  I agree that this sort of a visualization fits better with Loop's one-shot bolus paradigm than for OpenAPS SMB.  You're probably right that it'd be hard to adapt it to cover both, but at the same time the visualization is still useful (if not optimized) for the OpenAPS case, so I don't think it needs to be renamed: Predictions seems reasonable to me.

Reply:
@scottleibrand @MilosKozak When I looked at AndroidAPS data I only saw the IOB prediction curve, never the COB. For OpenAPS COB curve takes both Carbs and IOB into account (and is only present when carbs are present). But in what key is the corresponding prediction data for AndroidAPS stored?

Would it make sense to show COB curve if present, and IOB curve otherwise?

Reply:
I would fall back from COB if present, to UAM, then IOB.